### PR TITLE
#1642 add rootdir option

### PR DESCRIPTION
**Pull request by feuillemorte**
Added rootdir option.
How to test:

create folder with test environment, for example: /home/user/github/pytest
create folder with tests, for example: /home/user/rootest/tests/<tests_py>

enter test environment and run `pytest tests --rootdir=$HOME/rootest`

create pytest.in file in tests environment:
[pytest]
rootdir=$HOME/rootest

run tests

-------------
rootest
| - spoon.py
\ -tests
    \ - test_one.py (import spoon)

testruns:

`pytest tests --rootdir=$HOME/rootest`
`1 passed in 0.02 seconds`

### CODE REVIEW

File: _pytest/main.py [hunk 1]
```diff
@@ -283,6 +291,17 @@ def __init__(self, config):
         self.trace = config.trace.root.get("collection")
         self._norecursepatterns = config.getini("norecursedirs")
         self.startdir = py.path.local()
+
+        rootdir_ini = config.getini('rootdir')
```
Comment (nicoddemus) [inline _pytest/main.py]: This doesn't seem the proper place to override the `rootdir`, we should instead check if the option was passed in the `determine_setup` function, which is responsible to automatically discover the `rootdir`:

https://github.com/pytest-dev/pytest/blob/621374679b0cc9be4bbb76589fd94f20ecb9e0ec/_pytest/config.py#L1326-L1349

We should obtain the `rootdir` from the command line arguments in `Config.__init__` and pass it to this function, which should use the passed `rootdir` (if any) or go about discovering it like it does today if not passed on the command-line.

File: _pytest/main.py [hunk 2]
```diff
@@ -33,6 +33,9 @@ def pytest_addoption(parser):
     parser.addini("testpaths", "directories to search for tests when no files or directories are given in the "
                                "command line.",
                   type="args", default=[])
+    parser.addini("rootdir", "define root directory for tests. If this parameter defined command argument "
```
Comment (nicoddemus) [inline _pytest/main.py]: In general we should avoid adding both an ini option and command-line option for the same thing, because it is easy to change the value of an ini option in the command line (using `-o`) and one can change command-line options by setting `addopts` in an ini file, so it is a bit redundant.

In this case I think we will have to go with only a command-line option, because we determine the `rootdir` very early, even before we are setup to parse the `ini` file.

File: _pytest/main.py (hunk 1 above)
Comment (feuillemorte) [reply _pytest/main.py]: > We should obtain the rootdir from the command line arguments in Config.__init__

How can I obtain it?
command options are empty at this moment

`self.rootdir_cmd_arg = self.getoption('--rootdir')`
returns 
`ValueError: no option named '--rootdir'`
and there is a function `pytest_cmdline_parse` which executes after config.

`determine_setup` function executes in `_initini` and (as I think) ini files parsed before comand line. Am I right?

btw, how can I attach file from repo here like you?

Comment (nicoddemus) [reply _pytest/main.py]: I will have to think about how to obtain the command line option at that moment, but here's how you can post code snippets in comments on GitHub:

https://help.github.com/articles/creating-a-permanent-link-to-a-code-snippet/

Comment (feuillemorte) [reply _pytest/main.py]: https://github.com/pytest-dev/pytest/blob/621374679b0cc9be4bbb76589fd94f20ecb9e0ec/_pytest/config.py#L991
ns.rootdir is working and when I pass to 
https://github.com/pytest-dev/pytest/blob/621374679b0cc9be4bbb76589fd94f20ecb9e0ec/_pytest/config.py#L1326
some code like this:
```
    if rootdir_cmd_arg:
        rootdir_abs_path = py.path.local(rootdir_cmd_arg)
        if not os.path.isdir(str(rootdir_abs_path)):
            raise UsageError("Directory '{}' not found. Check your '--rootdir' option.".format(rootdir_abs_path))
        rootdir = rootdir_abs_path
    return rootdir, inifile, inicfg or {}
```

my rootdir was changed, but it's not working =) It try to find modules not in rootdir. Searching solution...

**Comment by feuillemorte** [comment]
`pytest tests --rootdir=invalid_folder`
not run tests and returns
`ERROR: Directory '/home/feuillemorte/github/pytest_fork/invalid_folder' not found. Check your '--rootdir' option.`

and I can't to write test on it because
```
result = testdir.runpytest("--rootdir=wrong_dir")
result.stdout.str()
```
returns empty string

Please, review this fix and write comments

File: _pytest/config.py [hunk 3]
```diff
@@ -990,11 +990,15 @@ def pytest_load_initial_conftests(self, early_config):
...
-        r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn)
+        rootdir = ns.rootdir if ns.rootdir else None
+        r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn, rootdir_cmd_arg=rootdir)
         self.rootdir, self.inifile, self.inicfg = r
         self._parser.extra_info['rootdir'] = self.rootdir
         self._parser.extra_info['inifile'] = self.inifile
         self.invocation_dir = py.path.local()
+        if ns.rootdir:
+            self.invocation_dir = self.rootdir
+            sys.path.append(str(self.rootdir))
```
Comment (nicoddemus) [inline _pytest/config.py]: This seems strange to me, why is this change required?

Comment (feuillemorte) [reply _pytest/config.py]: > what is the problem that you are encountering exactly

I want to write negative test and I can't match result. It's can't catch

> This seems strange to me, why is this change required?
```
            for path in sys.path:
                print(path)
```
```
/home/feuillemorte/github/pytest_fork/env/bin
/home/feuillemorte/github/pytest_fork/env/lib/python36.zip
/home/feuillemorte/github/pytest_fork/env/lib/python3.6
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/lib-dynload
/usr/lib64/python3.6
/usr/lib/python3.6
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages/pluggy-0.6.0-py3.6.egg
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages/attrs-17.4.0-py3.6.egg
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages/six-1.11.0-py3.6.egg
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages/py-1.5.2-py3.6.egg
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages/pytest-3.3.3.dev50+g83034bbd.d20180122-py3.6.egg
/home/feuillemorte/github/pytest_fork/env/lib/python3.6/site-packages
```

Python paths are not contains rootdir, so, I get an error:
`ModuleNotFoundError: No module named 'spoon'`

if it's not a good solution, please, give me an advice
Thank you!

Comment (nicoddemus) [reply _pytest/config.py]: > and I can't to write test on it because

If that's because `result.stdout.str()` is returning an empty string it might be that you have to look for the error in `result.stderr.str()` instead. 

Perhaps you can post the exact test you're trying and what you expected? That way I might be able to point the exact problem. 

> Python paths are not contain rootdir

Oh you mean in the test? If that's the case you can change the *test* instead to include the path by calling `testdir.syspathinsert()` at the start of the test.

Comment (feuillemorte) [reply _pytest/config.py]: > Oh you mean in the test

It's not only in test. It repeate when I test locally a rootdir option

Comment (nicoddemus) [reply _pytest/config.py]: But the `rootdir` is not added currently, so I don't think it should be added if `--rootdir` is passed in the command-line either.

Comment (feuillemorte) [reply _pytest/config.py]: I can check rootdir path and add to sys.path if it's not exist. Is it ok?

Comment (feuillemorte) [reply _pytest/config.py]: I mean if you have tests in another folder, for example, `/home/user/some_dir/tests` you don't have this folder in paths. So, you can't import some files by this path if you pass arg `--rootdir=/home/user/some_dir/tests` - you'll get an import error

Comment (nicoddemus) [reply _pytest/config.py]: @feuillemorte I see what you mean, but "automatically adding rootdir to sys.path" is a different issue than "set the rootdir on the command-line"; I don't think we should do both changes in the same PR and without some serious consideration.

Adding the `rootdir` to `sys.path` is not something to be done lightly because this can cause a lot of unexpected consequences, believe me. 😬

Comment (feuillemorte) [reply _pytest/config.py]: I see, thank you. But really, I don't know in which cases rootdir will be used because we can't use imports there (except system imports
I'll delete this line then :)

Comment (feuillemorte) [reply _pytest/config.py]: Only if import will be like 'from root.spoon...'
And in this case we can run like 'pytest root/tests' without rootdir option. Am I wrong?

Comment (nicoddemus) [reply _pytest/config.py]: The `rootdir` is used so pytest can have a reproducible node ids between runs in different computers. Here's an excerpt [from the docs](https://docs.pytest.org/en/latest/customize.html?highlight=rootdir#initialization-determining-rootdir-and-inifile):

> * Construct nodeids during collection; each test is assigned a unique nodeid which is rooted at the rootdir and takes in account full path, class name, function name and parametrization (if any).
> * Is used by plugins as a stable location to store project/test run specific information; for example, the internal cache plugin creates a .cache subdirectory in rootdir to store its cross-test run state.

So it is not really related to `sys.path` at all... but I understand why it might be confusing given its name.

**Comment by nicoddemus** [comment]
@feuillemorte could you please add a paragraph to [custommize](https://github.com/pytest-dev/pytest/blob/master/doc/en/customize.rst) mentioning this new option? Thanks!

File: _pytest/config.py (hunk 3 above)
Comment (feuillemorte) [reply _pytest/config.py]: I'll fix it asap )) I think today/tomorrow evening

**Comment by feuillemorte** [comment]
Fixed. Please, review

File: testing/test_session.py [hunk 4]
```diff
@@ -253,3 +253,28 @@ def pytest_sessionfinish():
     """)
     res = testdir.runpytest("--collect-only")
     assert res.ret == EXIT_NOTESTSCOLLECTED
+
+
+def test_rootdir_option_arg(testdir):
```
Comment (nicoddemus) [inline testing/test_session.py]: Please also add tests for absolute and paths containing environment variables.

File: testing/test_session.py [hunk 5]
```diff
@@ -253,3 +253,28 @@ def pytest_sessionfinish():
...
+    result.stdout.fnmatch_lines(["*AssertionError*"])
+
+    result = testdir.runpytest("--rootdir=root")
+    result.stdout.fnmatch_lines(["*1 passed*"])
+
+
+def test_rootdir_wrong_option_arg(testdir):
+    rootdir = testdir.mkdir("root")
+    testsdir = rootdir.mkdir("tests")
+    testsdir.join("test_one.py").write("def test_one():\n    assert 1")
```
Comment (nicoddemus) [inline testing/test_session.py]: This can be written as:

```python
testdir.makepyfile("""
    def test_one():
        assert 1
""")
``` 

Which is more readable IMO. This call will create a file `test_rootdir_wrong_option_arg.py` in the testdir directory (same name as the test), which is not really relevant for his this test.

Actually I don't even think the file or the directories is relevant for the test at all, so we can remove this preparation from this test.

File: _pytest/config.py [hunk 6]
```diff
@@ -990,11 +990,14 @@ def pytest_load_initial_conftests(self, early_config):
...
         ns, unknown_args = self._parser.parse_known_and_unknown_args(args, namespace=self.option.copy())
-        r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn)
+        rootdir = ns.rootdir if ns.rootdir else None
+        r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn, rootdir_cmd_arg=rootdir)
         self.rootdir, self.inifile, self.inicfg = r
         self._parser.extra_info['rootdir'] = self.rootdir
         self._parser.extra_info['inifile'] = self.inifile
         self.invocation_dir = py.path.local()
+        if ns.rootdir:
+            self.invocation_dir = self.rootdir
```
Comment (nicoddemus) [inline _pytest/config.py]: Why is this here? I'm not sure we should be changing the `invocation_dir`, it is supposed to contain the directory where pytest was invoked from, regardless of the `rootdir`.

File: testing/test_session.py (hunk 4 above)
Comment (feuillemorte) [reply testing/test_session.py]: paths containing environment variables not working in tests, but working when I test locally :(

```
testdir.runpytest("--rootdir=$HOME")

ERROR: Directory '/tmp/pytest-of-feuillemorte/pytest-31/test_rootdir_option_arg2/$HOME/root' not found. Check your '--rootdir' option.
```

Comment (nicoddemus) [reply testing/test_session.py]: You need to explicitly call `os.path.expandvars` to deal with variable expansion.

When you say "locally" I suppose you mean the command line? In that case is your shell doing the expansion, not pytest. 😉

File: testing/test_session.py (hunk 5 above)
Comment (feuillemorte) [reply testing/test_session.py]: It seems that I can't create file in root/tests folder by this code. It will created in the testdir folder. But tests runs fine, so, fixed =)

File: doc/en/customize.rst [hunk 7]
```diff
@@ -371,3 +371,9 @@ passed multiple times. The expected format is ``name=value``. For example::
 
 
 .. _`#3155`: https://github.com/pytest-dev/pytest/issues/3155
+
+.. confval:: rootdir
```
Comment (nicoddemus) [inline doc/en/customize.rst]: I just realized that this is not really a configuration variable (e.g goes into a `pytest.ini` file) but a command line option.

I think it makes more sense to add a paragraph to the `Initialization: determining rootdir and inifile` section in this same file instead.

File: _pytest/config.py [hunk 8]
```diff
@@ -991,7 +991,8 @@ def pytest_load_initial_conftests(self, early_config):
 
     def _initini(self, args):
         ns, unknown_args = self._parser.parse_known_and_unknown_args(args, namespace=self.option.copy())
-        r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn)
+        rootdir = ns.rootdir if ns.rootdir else None
+        r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn, rootdir_cmd_arg=rootdir)
```
Comment (nicoddemus) [inline _pytest/config.py]: This could be written as:

```python
r = determine_setup(ns.inifilename, ns.file_or_dir + unknown_args, warnfunc=self.warn, rootdir_cmd_arg=ns.rootdir or None)
```

Besides shorter, this doesn't left a `rootdir` local variable to be used by the unaware. 😁

File: _pytest/config.py [hunk 9]
```diff
@@ -1346,6 +1347,11 @@ def determine_setup(inifile, args, warnfunc=None):
                     is_fs_root = os.path.splitdrive(str(rootdir))[1] == '/'
                     if is_fs_root:
                         rootdir = ancestor
+    if rootdir_cmd_arg:
+        rootdir_abs_path = py.path.local(rootdir_cmd_arg)
```
Comment (nicoddemus) [inline _pytest/config.py]: You should call `os.path.expandvars(rootdir_cmd_arg)` here, not in the test; we want to make sure the functionality is available to users after all. 😉

File: testing/test_session.py [hunk 10]
```diff
@@ -253,3 +256,34 @@ def pytest_sessionfinish():
...
+
+    rootdir = testdir.mkdir("root")
+    rootdir.mkdir("tests")
+    testdir.makepyfile("""
+        import os
+        def test_one():
+            assert 1
+    """)
+
+    result = testdir.runpytest("--rootdir={}".format(os.path.expandvars(path)))
```
Comment (nicoddemus) [inline testing/test_session.py]: As I commented earlier we need to remove this `os.path.expandvars` call from here.

File: testing/test_session.py [hunk 11]
```diff
@@ -253,3 +256,34 @@ def pytest_sessionfinish():
...
     res = testdir.runpytest("--collect-only")
     assert res.ret == EXIT_NOTESTSCOLLECTED
+
+
+@pytest.mark.parametrize("path", ["root", "{relative}/root", "{environment}/root"])
+def test_rootdir_option_arg(testdir, path):
+    if 'relative' in path:
+        path = path.format(relative=os.getcwd())
+    if 'environment' in path:
+        os.environ['PY_ROOTDIR_PATH'] = os.getcwd()
```
Comment (nicoddemus) [inline testing/test_session.py]: Better to use `monkeypatch.setenv` instead so this variable is removed at the end of the test.

File: testing/test_session.py [hunk 12]
```diff
@@ -253,3 +256,34 @@ def pytest_sessionfinish():
     """)
     res = testdir.runpytest("--collect-only")
     assert res.ret == EXIT_NOTESTSCOLLECTED
+
+
+@pytest.mark.parametrize("path", ["root", "{relative}/root", "{environment}/root"])
+def test_rootdir_option_arg(testdir, path):
+    if 'relative' in path:
```
Comment (nicoddemus) [inline testing/test_session.py]: You can simplify the setup because applying a `format()` call to a string without replacements will yield the original string:

```python
def test_rootdir_option_arg(testdir, path):
    path = path.format(relative=str(testdir.tmpdir), 
                       environment='$PY_ROOTDIR_PATH')
```

File: testing/test_session.py [hunk 13]
```diff
@@ -253,3 +256,34 @@ def pytest_sessionfinish():
...
+        path = path.format(relative=os.getcwd())
```
Comment (nicoddemus) [inline testing/test_session.py]: You should not be using `os.getcwd()` here because this might leave some files after the run (the `.pytest_cache` directory for example). Use `str(testdir.tmpdir)` instead.

File: testing/test_session.py [hunk 14]
```diff
@@ -253,3 +256,34 @@ def pytest_sessionfinish():
...
+    rootdir = testdir.mkdir("root")
+    rootdir.mkdir("tests")
+    testdir.makepyfile("""
+        import os
+        def test_one():
+            assert 1
+    """)
+
+    result = testdir.runpytest("--rootdir={}".format(os.path.expandvars(path)))
+    result.stdout.fnmatch_lines(['*rootdir: {}/root, inifile:*'.format(os.getcwd()), "*1 passed*"])
```
Comment (nicoddemus) [inline testing/test_session.py]: Ditto here about `os.getcwd()`
